File: CiaiControllerSDK/Config/DriverConfig.cs
using System;
using System.Collections.Generic;
using YamlDotNet.Serialization;

namespace CiaiControllerSDK.Config
{
    /// <summary>
    /// 驱动配置模型，对应application.yml文件结构
    /// 参照 IncubatorController 的 Spring Boot SSL 配置
    /// </summary>
    public class DriverConfig
    {
        /// <summary>
        /// 服务器配置
        /// </summary>
        public ServerConfig Server { get; set; }

        /// <summary>
        /// 回调配置
        /// </summary>
        public CallbackConfig Callback { get; set; }

        /// <summary>
        /// 设备配置
        /// </summary>
        public DeviceConfigSection Device { get; set; }
    }

    /// <summary>
    /// 服务器配置
    /// </summary>
    public class ServerConfig
    {
        public int MaxConcurrentRequests { get; set; } = 100;
        public int MaxRequestBodyBytes { get; set; } = 1024 * 1024;
        public int FunctionQueueCapacity { get; set; } = 100;
        public int IdempotencyCapacity { get; set; } = 10000;
        public int ShutdownTimeoutMs { get; set; } = 30000;
        public bool AllowReplaceCertificateBinding { get; set; }
        /// <summary>
        /// 监听端口
        /// </summary>
        public int Port { get; set; } = 443;

        /// <summary>
        /// 主机名
        /// </summary>
        public string Host { get; set; } = "localhost";

        /// <summary>
        /// 是否使用HTTPS，默认true
        /// </summary>
        public bool UseHttps { get; set; } = true;

        /// <summary>
        /// 证书配置（密钥库）
        /// 对应 Spring Boot: server.ssl.key-store-*
        /// </summary>
        public CertificateConfig Certificate { get; set; }

        /// <summary>
        /// 信任库配置
        /// 对应 Spring Boot: server.ssl.trust-store-*
        /// </summary>
        public TrustStoreConfig TrustStore { get; set; }

        /// <summary>
        /// SSL/TLS 配置
        /// 对应 Spring Boot: server.ssl.*
        /// </summary>
        public SslConfig Ssl { get; set; }

        /// <summary>
        /// 客户端认证配置
        /// 对应 Spring Boot: server.ssl.client-auth
        /// </summary>
        public ClientAuthConfig ClientAuth { get; set; }
    }

    /// <summary>
    /// 证书配置（密钥库）
    /// 对应 Spring Boot: server.ssl.key-store-*
    /// </summary>
    public class CertificateConfig
    {
        /// <summary>
        /// 证书文件路径
        /// </summary>
        public string Path { get; set; }

        /// <summary>
        /// 证书密码
        /// </summary>
        public string Password { get; set; }

        /// <summary>
        /// 密钥库类型，默认 PKCS12
        /// </summary>
        public string Type { get; set; } = "PKCS12";

        /// <summary>
        /// 密钥别名
        /// </summary>
        public string Alias { get; set; }
    }

    /// <summary>
    /// 信任库配置
    /// 对应 Spring Boot: server.ssl.trust-store-*
    /// </summary>
    public class TrustStoreConfig
    {
        /// <summary>
        /// 信任库文件路径
        /// </summary>
        public string Path { get; set; }

        /// <summary>
        /// 信任库密码
        /// </summary>
        public string Password { get; set; }

        /// <summary>
        /// 信任库类型，默认 PKCS12
        /// </summary>
        public string Type { get; set; } = "PKCS12";
    }

    /// <summary>
    /// SSL/TLS 配置
    /// 对应 Spring Boot: server.ssl.protocol, ciphers, enabled-protocols
    /// </summary>
    public class SslConfig
    {
        /// <summary>
        /// SSL协议版本，默认 TLSv1.3
        /// 对应 Spring Boot: server.ssl.protocol
        /// </summary>
        public string Protocol { get; set; } = "TLSv1.3";

        /// <summary>
        /// 启用的协议列表
        /// 对应 Spring Boot: server.ssl.enabled-protocols
        /// </summary>
        public string[] EnabledProtocols { get; set; }

        /// <summary>
        /// 加密套件列表
        /// 对应 Spring Boot: server.ssl.ciphers
        /// </summary>
        public string[] Ciphers { get; set; }
    }

    /// <summary>
    /// 客户端认证配置
    /// 对应 Spring Boot: server.ssl.client-auth
    /// </summary>
    public class ClientAuthConfig
    {
        /// <summary>
        /// 客户端认证模式: need, want, none
        /// 对应 Spring Boot: server.ssl.client-auth
        /// </summary>
        public string Mode { get; set; } = "need";

        /// <summary>
        /// 是否启用客户端证书验证
        /// </summary>
        public bool Enabled { get; set; } = true;

        /// <summary>
        /// 受信任的客户端证书指纹列表
        /// </summary>
        public string[] TrustedThumbprints { get; set; }

        /// <summary>
        /// 受信任的CA指纹列表
        /// </summary>
        public string[] TrustedIssuers { get; set; }
    }

    /// <summary>
    /// 回调配置
    /// </summary>
    public class CallbackConfig
    {
        /// <summary>
        /// 回调URL
        /// </summary>
        public string Url { get; set; }

        /// <summary>
        /// 回调超时时间（毫秒）
        /// </summary>
        public int TimeoutMs { get; set; } = 30000;

        /// <summary>
        /// 是否启用回调
        /// </summary>
        public bool Enabled { get; set; } = true;
    }

    /// <summary>
    /// 设备配置部分
    /// </summary>
    public class DeviceConfigSection
    {
        /// <summary>
        /// 设备ID
        /// </summary>
        public string DeviceId { get; set; }

        /// <summary>
        /// 通信类型: TCP, HTTP, Serial, DLL
        /// </summary>
        public string CommunicationType { get; set; }

        /// <summary>
        /// TCP配置
        /// </summary>
        public TcpConfig Tcp { get; set; }

        /// <summary>
        /// HTTP配置
        /// </summary>
        public HttpConfig Http { get; set; }

        /// <summary>
        /// 串口配置
        /// </summary>
        public SerialConfig Serial { get; set; }

        /// <summary>多个底层连接；键就是驱动代码中使用的连接名。</summary>
        public Dictionary<string, ConnectionConfig> Connections { get; set; } = new();

        /// <summary>
        /// 底层设备/DLL调用并发数；不影响上位系统的业务并行能力。
        /// </summary>
        public int DeviceCallResources { get; set; } = 1;

        /// <summary>
        /// 等待底层设备调用资源超时（毫秒）。
        /// </summary>
        public int DeviceCallTimeoutMs { get; set; } = 30000;

        /// <summary>
        /// 驱动自定义配置。建议在YAML中放到 device.settings 下。
        /// device 下未被SDK识别的字段也会自动合并到这里。
        /// </summary>
        public Dictionary<string, object> Settings { get; set; } = new();

        [YamlIgnore]
        public Dictionary<string, object> AdditionalSettings { get; set; } = new();
    }

    /// <summary>
    /// TCP配置
    /// </summary>
    public class TcpConfig
    {
        public string Host { get; set; }
        public int Port { get; set; }
        public int TimeoutMs { get; set; } = 5000;
        public int? ConnectTimeoutMs { get; set; }
        public int? ReadTimeoutMs { get; set; }
        public int? WriteTimeoutMs { get; set; }
    }

    /// <summary>
    /// HTTP配置
    /// </summary>
    public class HttpConfig
    {
        public string BaseUrl { get; set; }
        public int TimeoutMs { get; set; } = 30000;
    }

    /// <summary>
    /// 串口配置
    /// </summary>
    public class SerialConfig
    {
        public string Port { get; set; }
        public int BaudRate { get; set; } = 9600;
        public int DataBits { get; set; } = 8;
        public double StopBits { get; set; } = 1;
        public string Parity { get; set; } = "none";
        public int TimeoutMs { get; set; } = 5000;
        public int? ReadTimeoutMs { get; set; }
        public int? WriteTimeoutMs { get; set; }
        public string Encoding { get; set; } = "utf-8";
        public string FlowControl { get; set; } = "none";
        public bool DtrEnable { get; set; }
        public bool RtsEnable { get; set; }
        public bool DiscardInputBeforeWrite { get; set; }
    }

    /// <summary>命名连接的声明式配置。</summary>
    public class ConnectionConfig
    {
        public string Type { get; set; }
        public bool Default { get; set; }
        public bool Required { get; set; } = true;
        public bool ConnectOnStart { get; set; } = true;
        public string Host { get; set; }
        public int Port { get; set; }
        public string BaseUrl { get; set; }
        public string SerialPort { get; set; }
        public int BaudRate { get; set; } = 9600;
        public int DataBits { get; set; } = 8;
        public double StopBits { get; set; } = 1;
        public string Parity { get; set; } = "none";
        public string Encoding { get; set; } = "utf-8";
        public string FlowControl { get; set; } = "none";
        public bool DtrEnable { get; set; }
        public bool RtsEnable { get; set; }
        public bool DiscardInputBeforeWrite { get; set; }
        public int ConnectTimeoutMs { get; set; } = 5000;
        public int ReadTimeoutMs { get; set; } = 10000;
        public int WriteTimeoutMs { get; set; } = 10000;
        public int ResourceWaitTimeoutMs { get; set; } = 30000;
        public int MaxConcurrency { get; set; } = 1;
        public string ResourceGroup { get; set; }
        public int RetryCount { get; set; }
        public int RetryDelayMs { get; set; } = 200;
        public double RetryBackoff { get; set; } = 2;
        public string Executable { get; set; }
        public List<string> Arguments { get; set; } = new();
        public string WorkingDirectory { get; set; }
        public string Architecture { get; set; } = "auto";
        public string Framework { get; set; }
        public string ApartmentState { get; set; } = "MTA";
        public int ShutdownTimeoutMs { get; set; } = 5000;
        public Dictionary<string, string> Environment { get; set; } = new();
        public Dictionary<string, string> Headers { get; set; } = new();
        public Dictionary<string, object> Settings { get; set; } = new();
    }
}

File: CiaiControllerSDKForJava/src/main/java/com/ciai/controller/sdk/config/DriverConfig.java
package com.ciai.controller.sdk.config;

import com.ciai.controller.sdk.core.ConnectionConfiguration;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 驱动配置模型
 */
public class DriverConfig {

    private ServerConfig server;
    private CallbackConfig callback;
    private DeviceConfigSection device;
    private transient String sourceDirectory;

    public String getSourceDirectory() {
        return sourceDirectory;
    }

    void setSourceDirectory(String sourceDirectory) {
        this.sourceDirectory = sourceDirectory;
    }

    public ServerConfig getServer() {
        return server;
    }

    public void setServer(ServerConfig server) {
        this.server = server;
    }

    public CallbackConfig getCallback() {
        return callback;
    }

    public void setCallback(CallbackConfig callback) {
        this.callback = callback;
    }

    public DeviceConfigSection getDevice() {
        return device;
    }

    public void setDevice(DeviceConfigSection device) {
        this.device = device;
    }

    /**
     * 服务器配置
     * 参照 IncubatorController 的 Spring Boot SSL 配置
     */
    public static class ServerConfig {
        private int port = 8080;
        private String host = "localhost";
        private boolean useHttps = false;
        private CertificateConfig certificate;
        private TrustStoreConfig trustStore;  // 信任库配置（新增）
        private ClientAuthConfig clientAuth;
        private SslConfig ssl;  // SSL/TLS 配置（新增）
        private int maxConcurrentRequests = 100;
        private int maxRequestBodyBytes = 1024 * 1024;
        private int functionQueueCapacity = 100;
        private int idempotencyCapacity = 10000;
        private int shutdownTimeoutMs = 30000;
        private boolean allowReplaceCertificateBinding;

        public int getPort() {
            return port;
        }

        public void setPort(int port) {
            this.port = port;
        }

        public String getHost() {
            return host;
        }

        public void setHost(String host) {
            this.host = host;
        }

        public boolean isUseHttps() {
            return useHttps;
        }

        public void setUseHttps(boolean useHttps) {
            this.useHttps = useHttps;
        }

        public CertificateConfig getCertificate() {
            return certificate;
        }

        public void setCertificate(CertificateConfig certificate) {
            this.certificate = certificate;
        }

        public TrustStoreConfig getTrustStore() {
            return trustStore;
        }

        public void setTrustStore(TrustStoreConfig trustStore) {
            this.trustStore = trustStore;
        }

        public ClientAuthConfig getClientAuth() {
            return clientAuth;
        }

        public void setClientAuth(ClientAuthConfig clientAuth) {
            this.clientAuth = clientAuth;
        }

        public SslConfig getSsl() {
            return ssl;
        }

        public void setSsl(SslConfig ssl) {
            this.ssl = ssl;
        }

        public int getMaxConcurrentRequests() { return maxConcurrentRequests; }
        public void setMaxConcurrentRequests(int value) { this.maxConcurrentRequests = value; }
        public int getMaxRequestBodyBytes() { return maxRequestBodyBytes; }
        public void setMaxRequestBodyBytes(int value) { this.maxRequestBodyBytes = value; }
        public int getFunctionQueueCapacity() { return functionQueueCapacity; }
        public void setFunctionQueueCapacity(int value) { this.functionQueueCapacity = value; }
        public int getIdempotencyCapacity() { return idempotencyCapacity; }
        public void setIdempotencyCapacity(int value) { this.idempotencyCapacity = value; }
        public int getShutdownTimeoutMs() { return shutdownTimeoutMs; }
        public void setShutdownTimeoutMs(int value) { this.shutdownTimeoutMs = value; }
        public boolean isAllowReplaceCertificateBinding() { return allowReplaceCertificateBinding; }
        public void setAllowReplaceCertificateBinding(boolean value) { this.allowReplaceCertificateBinding = value; }
    }

    /**
     * 证书配置
     * 对应 Spring Boot: server.ssl.key-store-*
     */
    public static class CertificateConfig {
        private String path;
        private String password;
        private String type = "PKCS12";  // 密钥库类型
        private String alias;  // 密钥别名

        public String getPath() {
            return path;
        }

        public void setPath(String path) {
            this.path = path;
        }

        public String getPassword() {
            return password;
        }

        public void setPassword(String password) {
            this.password = password;
        }

        public String getType() {
            return type;
        }

        public void setType(String type) {
            this.type = type;
        }

        public String getAlias() {
            return alias;
        }

        public void setAlias(String alias) {
            this.alias = alias;
        }
    }

    /**
     * 信任库配置
     * 对应 Spring Boot: server.ssl.trust-store-*
     */
    public static class TrustStoreConfig {
        private String path;
        private String password;
        private String type = "PKCS12";

        public String getPath() {
            return path;
        }

        public void setPath(String path) {
            this.path = path;
        }

        public String getPassword() {
            return password;
        }

        public void setPassword(String password) {
            this.password = password;
        }

        public String getType() {
            return type;
        }

        public void setType(String type) {
            this.type = type;
        }
    }

    /**
     * SSL/TLS 配置
     * 对应 Spring Boot: server.ssl.protocol, server.ssl.ciphers, server.ssl.enabled-protocols
     */
    public static class SslConfig {
        private String protocol = "TLSv1.2";  // 对应 server.ssl.protocol
        private List<String> enabledProtocols;  // 对应 server.ssl.enabled-protocols
        private List<String> ciphers;  // 对应 server.ssl.ciphers

        public String getProtocol() {
            return protocol;
        }

        public void setProtocol(String protocol) {
            this.protocol = protocol;
        }

        public List<String> getEnabledProtocols() {
            return enabledProtocols;
        }

        public void setEnabledProtocols(List<String> enabledProtocols) {
            this.enabledProtocols = enabledProtocols;
        }

        public List<String> getCiphers() {
            return ciphers;
        }

        public void setCiphers(List<String> ciphers) {
            this.ciphers = ciphers;
        }
    }

    /**
     * 客户端认证配置
     * 对应 Spring Boot: server.ssl.client-auth
     */
    public static class ClientAuthConfig {
        private String mode = "need";  // need, want, none（对应 Spring Boot: server.ssl.client-auth）
        private boolean enabled = true;
        private List<String> trustedThumbprints;

        public String getMode() {
            return mode;
        }

        public void setMode(String mode) {
            this.mode = mode;
        }

        public boolean isEnabled() {
            return enabled;
        }

        public void setEnabled(boolean enabled) {
            this.enabled = enabled;
        }

        public List<String> getTrustedThumbprints() {
            return trustedThumbprints;
        }

        public void setTrustedThumbprints(List<String> trustedThumbprints) {
            this.trustedThumbprints = trustedThumbprints;
        }
    }

    /**
     * 回调配置
     */
    public static class CallbackConfig {
        private String url;
        private int timeoutMs = 30000;
        private boolean enabled = true;

        public String getUrl() {
            return url;
        }

        public void setUrl(String url) {
            this.url = url;
        }

        public int getTimeoutMs() {
            return timeoutMs;
        }

        public void setTimeoutMs(int timeoutMs) {
            this.timeoutMs = timeoutMs;
        }

        public boolean isEnabled() {
            return enabled;
        }

        public void setEnabled(boolean enabled) {
            this.enabled = enabled;
        }
    }

    /**
     * 设备配置部分
     */
    public static class DeviceConfigSection {
        private String deviceId;
        private String communicationType;
        private int deviceCallResources = 1;
        private int deviceCallTimeoutMs = 30000;
        private TcpConfig tcp;
        private HttpConfig http;
        private SerialConfig serial;
        private Map<String, ConnectionConfiguration> connections = new LinkedHashMap<>();
        private Map<String, Object> settings = new LinkedHashMap<>();

        public String getDeviceId() {
            return deviceId;
        }

        public void setDeviceId(String deviceId) {
            this.deviceId = deviceId;
        }

        public String getCommunicationType() {
            return communicationType;
        }

        public void setCommunicationType(String communicationType) {
            this.communicationType = communicationType;
        }

        public int getDeviceCallResources() {
            return deviceCallResources;
        }

        public void setDeviceCallResources(int deviceCallResources) {
            this.deviceCallResources = deviceCallResources;
        }

        public int getDeviceCallTimeoutMs() {
            return deviceCallTimeoutMs;
        }

        public void setDeviceCallTimeoutMs(int deviceCallTimeoutMs) {
            this.deviceCallTimeoutMs = deviceCallTimeoutMs;
        }

        public TcpConfig getTcp() {
            return tcp;
        }

        public void setTcp(TcpConfig tcp) {
            this.tcp = tcp;
        }

        public HttpConfig getHttp() {
            return http;
        }

        public void setHttp(HttpConfig http) {
            this.http = http;
        }

        public SerialConfig getSerial() {
            return serial;
        }

        public void setSerial(SerialConfig serial) {
            this.serial = serial;
        }

        public Map<String, ConnectionConfiguration> getConnections() { return connections; }
        public void setConnections(Map<String, ConnectionConfiguration> value) {
            connections = value == null ? new LinkedHashMap<String, ConnectionConfiguration>() : value;
        }

        public Map<String, Object> getSettings() {
            return settings;
        }

        public void setSettings(Map<String, Object> settings) {
            this.settings = settings == null ? new LinkedHashMap<>() : settings;
        }
    }

    /**
     * TCP配置
     */
    public static class TcpConfig {
        private String host;
        private int port;
        private int timeoutMs = 5000;
        private Integer connectTimeoutMs;
        private Integer readTimeoutMs;
        private Integer writeTimeoutMs;

        public String getHost() {
            return host;
        }

        public void setHost(String host) {
            this.host = host;
        }

        public int getPort() {
            return port;
        }

        public void setPort(int port) {
            this.port = port;
        }

        public int getTimeoutMs() {
            return timeoutMs;
        }

        public void setTimeoutMs(int timeoutMs) {
            this.timeoutMs = timeoutMs;
        }

        public Integer getConnectTimeoutMs() { return connectTimeoutMs; }
        public void setConnectTimeoutMs(Integer value) { this.connectTimeoutMs = value; }
        public Integer getReadTimeoutMs() { return readTimeoutMs; }
        public void setReadTimeoutMs(Integer value) { this.readTimeoutMs = value; }
        public Integer getWriteTimeoutMs() { return writeTimeoutMs; }
        public void setWriteTimeoutMs(Integer value) { this.writeTimeoutMs = value; }
    }

    /**
     * HTTP配置
     */
    public static class HttpConfig {
        private String baseUrl;
        private int timeoutMs = 30000;

        public String getBaseUrl() {
            return baseUrl;
        }

        public void setBaseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
        }

        public int getTimeoutMs() {
            return timeoutMs;
        }

        public void setTimeoutMs(int timeoutMs) {
            this.timeoutMs = timeoutMs;
        }
    }

    /**
     * 串口配置
     */
    public static class SerialConfig {
        private String port;
        private int baudRate = 9600;
        private int dataBits = 8;
        private double stopBits = 1;
        private String parity = "none";
        private int timeoutMs = 5000;
        private Integer readTimeoutMs;
        private Integer writeTimeoutMs;
        private String encoding = "utf-8";
        private String flowControl = "none";
        private boolean dtrEnable;
        private boolean rtsEnable;
        private boolean discardInputBeforeWrite;

        public String getPort() {
            return port;
        }

        public void setPort(String port) {
            this.port = port;
        }

        public int getBaudRate() {
            return baudRate;
        }

        public void setBaudRate(int baudRate) {
            this.baudRate = baudRate;
        }

        public int getDataBits() {
            return dataBits;
        }

        public void setDataBits(int dataBits) {
            this.dataBits = dataBits;
        }

        public double getStopBits() {
            return stopBits;
        }

        public void setStopBits(double stopBits) {
            this.stopBits = stopBits;
        }

        public String getParity() {
            return parity;
        }

        public void setParity(String parity) {
            this.parity = parity;
        }

        public int getTimeoutMs() { return timeoutMs; }
        public void setTimeoutMs(int timeoutMs) { this.timeoutMs = timeoutMs; }
        public Integer getReadTimeoutMs() { return readTimeoutMs; }
        public void setReadTimeoutMs(Integer value) { this.readTimeoutMs = value; }
        public Integer getWriteTimeoutMs() { return writeTimeoutMs; }
        public void setWriteTimeoutMs(Integer value) { this.writeTimeoutMs = value; }
        public String getEncoding() { return encoding; }
        public void setEncoding(String encoding) { this.encoding = encoding; }
        public String getFlowControl() { return flowControl; }
        public void setFlowControl(String value) { flowControl = value; }
        public boolean isDtrEnable() { return dtrEnable; }
        public void setDtrEnable(boolean value) { dtrEnable = value; }
        public boolean isRtsEnable() { return rtsEnable; }
        public void setRtsEnable(boolean value) { rtsEnable = value; }
        public boolean isDiscardInputBeforeWrite() { return discardInputBeforeWrite; }
        public void setDiscardInputBeforeWrite(boolean value) { discardInputBeforeWrite = value; }
    }
}
